stdlib/public/runtime/CrashHandlerMacOS.cpp

```diff
@@ -0,0 +1,462 @@
+// Reporter, which is a no-no.
+//
+//===----------------------------------------------------------------------===//
+
+#ifdef __APPLE__
+
+#include <TargetConditionals.h>
+
+#if TARGET_OS_OSX
+
+#include <mach/mach.h>
+#include <mach/task.h>
+#include <mach/thread_act.h>
+
+#include <sys/mman.h>
+#include <sys/ucontext.h>
+#include <sys/wait.h>
+
+#include <os/lock.h>
+
+#include <errno.h>
+#include <signal.h>
+#include <spawn.h>
+#include <unistd.h>
+
+#include "swift/Runtime/Backtrace.h"
+
+#include <cstring>
+
+#ifndef lengthof
+#define lengthof(x)     (sizeof(x) / sizeof(x[0]))
+#endif
+
+using namespace swift::runtime::backtrace;
+
+namespace {
+
+void handle_fatal_signal(int signum, siginfo_t *pinfo, void *uctx);
+void suspend_other_threads();
+void resume_other_threads();
+bool run_backtracer(void);
+
+swift::CrashInfo crashInfo;
+
+os_unfair_lock crashLock = OS_UNFAIR_LOCK_INIT;
+
+const int signalsToHandle[] = {
+  SIGQUIT,
+  SIGABRT,
+  SIGBUS,
+  SIGFPE,
+  SIGILL,
+  SIGSEGV,
+  SIGTRAP
+};
+
+} // namespace
+
+namespace swift {
+namespace runtime {
+namespace backtrace {
+
+SWIFT_RUNTIME_STDLIB_INTERNAL int
+_swift_installCrashHandler()
+{
+  stack_t ss;
+
+  // See if an alternate signal stack already exists
+  if (sigaltstack(NULL, &ss) < 0)
+    return errno;
+
+  if (ss.ss_sp == 0) {
+    // No, so set one up
+    ss.ss_flags = 0;
+    ss.ss_size = SIGSTKSZ;
+    ss.ss_sp = mmap(0, ss.ss_size, PROT_READ | PROT_WRITE,
+                    MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    if (ss.ss_sp == MAP_FAILED)
+      return errno;
+
+    if (sigaltstack(&ss, NULL) < 0)
+      return errno;
+  }
+
+  // Now register signal handlers
+  struct sigaction sa;
+
+  sigfillset(&sa.sa_mask);
+  for (unsigned n = 0; n < lengthof(signalsToHandle); ++n) {
+    sigdelset(&sa.sa_mask, signalsToHandle[n]);
+  }
+
+  sa.sa_handler = NULL;
+  sa.sa_flags = SA_ONSTACK | SA_SIGINFO | SA_NODEFER;
+  sa.sa_sigaction = handle_fatal_signal;
+
+  for (unsigned n = 0; n < lengthof(signalsToHandle); ++n) {
+    struct sigaction osa;
+
+    // See if a signal handler for this signal is already installed
+    if (sigaction(signalsToHandle[n], NULL, &osa) < 0)
+      return errno;
+
+    if (osa.sa_handler == SIG_DFL) {
+      // No, so install ours
+      if (sigaction(signalsToHandle[n], &sa, NULL) < 0)
+        return errno;
+    }
+  }
+
+  return 0;
+}
+
+} // namespace backtrace
+} // namespace runtime
+} // namespace swift
+
+namespace {
+
+void
+suspend_other_threads()
+{
+  os_unfair_lock_lock(&crashLock);
+
+  thread_t self = mach_thread_self();
+  thread_act_array_t threads;
+  mach_msg_type_number_t count = 0;
+
+  kern_return_t kr = task_threads(mach_task_self(), &threads, &count);
+
+  if (kr != KERN_SUCCESS)
+    return;
+
+  for (unsigned n = 0; n < count; ++n) {
+    if (threads[n] == self)
+      continue;
+
+    // Ignore the results of these two; if they fail there's nothing we can do
+    (void)thread_resume(threads[n]);
```

Comment:
`thread_suspend`, I presume.

Reply:
Indeed. I don't know how I managed that. Good spot.